File: amber/typescript/src/incoming.ts
import {
  AmberTransitionError,
  AmberValidationError,
  Provenance,
} from "./provenance.js";
import { ProvenanceContext } from "./context.js";

export const MAX_INCOMING_JSON_BYTES = 16 * 1024;

export type IncomingPolicy = "reject" | "ignore";

/** Apply an application or deployment trust rule after structural decoding. */
export type IncomingValidator = (provenance: Provenance) => void;

export type IncomingInspection = {
  readonly provenance?: Provenance;
  readonly present: boolean;
};

export type IncomingContextResult = {
  readonly context: ProvenanceContext;
  readonly present: boolean;
};

function validatePolicy(policy: IncomingPolicy): void {
  if (policy !== "reject" && policy !== "ignore") {
    throw new AmberTransitionError(`unsupported incoming policy ${String(policy)}`);
  }
}

function absent(): IncomingInspection {
  return { present: false };
}

function invalid(policy: IncomingPolicy, error: unknown): IncomingInspection {
  if (policy === "ignore") {
    return absent();
  }
  if (error instanceof AmberValidationError) {
    throw error;
  }
  throw new AmberValidationError(String(error));
}

/** Inspect incoming JSON without installing it into a context. */
export function inspectIncomingJSON(
  input: string | null | undefined,
  policy: IncomingPolicy,
): IncomingInspection {
	return inspectIncomingJSONWithValidator(input, policy);
}

/** Validate incoming JSON and apply an optional trust validator. */
export function inspectIncomingJSONWithValidator(
  input: string | null | undefined,
  policy: IncomingPolicy,
  validator?: IncomingValidator,
): IncomingInspection {
	validatePolicy(policy);
  if (input === undefined || input === null || input.length === 0) {
    return absent();
  }
  if (typeof input !== "string") {
    return invalid(policy, new Error("incoming JSON must be a string"));
  }
  if (new TextEncoder().encode(input).byteLength > MAX_INCOMING_JSON_BYTES) {
    return invalid(
      policy,
      new Error(`incoming JSON exceeds ${MAX_INCOMING_JSON_BYTES} bytes`),
    );
	}
	try {
		const provenance = Provenance.fromJSON(input);
		if (validator !== undefined) {
			try {
				validator(provenance);
			} catch (error) {
				return invalid(policy, new Error(`incoming validation failed: ${String(error)}`));
			}
		}
		return { provenance, present: true };
	} catch (error) {
		return invalid(policy, error);
  }
}

/**
 * Inspect and, when accepted, install incoming JSON in a derived context.
 * Absent or ignored input returns the original context unchanged.
 */
export function withIncomingJSON(
  context: ProvenanceContext,
  input: string | null | undefined,
  policy: IncomingPolicy,
): IncomingContextResult {
	return withIncomingJSONWithValidator(context, input, policy);
}

/** Install incoming JSON only after an optional trust validator accepts it. */
export function withIncomingJSONWithValidator(
  context: ProvenanceContext,
  input: string | null | undefined,
  policy: IncomingPolicy,
  validator?: IncomingValidator,
): IncomingContextResult {
	if (!(context instanceof ProvenanceContext)) {
		throw new AmberTransitionError("context must be a ProvenanceContext instance");
	}
	const inspected = inspectIncomingJSONWithValidator(input, policy, validator);
  if (!inspected.present || inspected.provenance === undefined) {
    return { context, present: false };
  }
  return {
    context: context.withProvenance(inspected.provenance),
    present: true,
  };
}

File: amber/go/incoming.go
package amber

import (
	"context"
	"fmt"
)

const MaxIncomingJSONBytes = 16 * 1024

// IncomingPolicy controls how an adapter handles malformed incoming data.
type IncomingPolicy string

const (
	IncomingReject IncomingPolicy = "reject"
	IncomingIgnore IncomingPolicy = "ignore"
)

func validateIncomingPolicy(policy IncomingPolicy) error {
	if policy != IncomingReject && policy != IncomingIgnore {
		return fmt.Errorf("%w: unsupported incoming policy %q", ErrInvalidTransition, policy)
	}
	return nil
}

// InspectIncomingJSON validates an incoming JSON value without installing it
// into any context. Empty input is absent; non-empty invalid input is either
// rejected or treated as absent according to policy.
func InspectIncomingJSON(data []byte, policy IncomingPolicy) (Provenance, bool, error) {
	if err := validateIncomingPolicy(policy); err != nil {
		return Provenance{}, false, err
	}
	if len(data) == 0 {
		return Provenance{}, false, nil
	}
	if len(data) > MaxIncomingJSONBytes {
		return incomingFailure(policy, fmt.Errorf("incoming JSON exceeds %d bytes", MaxIncomingJSONBytes))
	}

	provenance, err := FromJSON(data)
	if err != nil {
		return incomingFailure(policy, err)
	}
	return provenance, true, nil
}

// WithIncomingJSON inspects and, when accepted, installs an incoming value in
// a derived context. Absent or ignored input returns the original context.
func WithIncomingJSON(ctx context.Context, data []byte, policy IncomingPolicy) (context.Context, bool, error) {
	if ctx == nil {
		return nil, false, fmt.Errorf("%w: context cannot be nil", ErrInvalidTransition)
	}
	provenance, present, err := InspectIncomingJSON(data, policy)
	if err != nil || !present {
		return ctx, present, err
	}
	installed, err := WithProvenance(ctx, provenance)
	if err != nil {
		return ctx, false, err
	}
	return installed, true, nil
}

func incomingFailure(policy IncomingPolicy, err error) (Provenance, bool, error) {
	if policy == IncomingIgnore {
		return Provenance{}, false, nil
	}
	return Provenance{}, false, fmt.Errorf("%w: %v", ErrInvalidProvenance, err)
}
